carmajava/signal_plugin/src/main/java/gov/dot/fhwa/saxton/carma/signal_plugin/ead/SimpleNCVMotionPredictor.java

```diff
@@ -51,57 +33,37 @@
 
     List<RoutePointStamped> projection = new LinkedList<>();
 
-    // Return an empty list if provided with less than 2 points
-    if (objTrajectory.size() < 2) {
+    // Return an empty list if provided with an empty list
+    if (objTrajectory.isEmpty()) {
       return projection; 
     }
 
-    double sumDist = 0; // Since we are using distance steps the independent variable will be distance
-    double sumTime = 0; // Since we are using distance steps the dependent variable will be time
-    double sumSqrDist = 0;
-    double sumDistxTime = 0;
-
-    // Calculate sums for regression
-    for (RoadwayObstacle n: objTrajectory) {
-      
-      double d = n.getDownTrack();
-      double t = n.getObject().getHeader().getStamp().toSeconds();
-      double d_sqr = d * d;
-      double td = t * d; 
-
-      sumDist += d;
-      sumTime += t;
-      sumSqrDist += d_sqr;
-      sumDistxTime += td; 
-    }
-
-    // Compute intercept and slope
-    double count = objTrajectory.size();
-    double denominator = (count * sumSqrDist - sumDist * sumDist);
-
-    double intercept = (sumTime * sumSqrDist - sumDist * sumDistxTime) / denominator;
-    double slope = (count * sumDistxTime - sumDist * sumTime) / denominator;
-
-    // Generate projection using regression model
+    // Calculate average speed from historical data
+    double averageSpeed = objTrajectory.stream()
+    		                       .mapToDouble(a -> a.getObject().getVelocity().getTwist().getLinear().getX())
+    		                       .average()
+    		                       .getAsDouble();
+    
+    // Generate projection using average speed
     // Set endtime as the last timestamp in the provided history plus the time duration 
     double startDist = objTrajectory.get(objTrajectory.size() - 1).getDownTrack();
     double d = startDist;
     double t = objTrajectory.get(objTrajectory.size() - 1).getObject().getHeader().getStamp().toSeconds();
     double endTime = timeDuration + t;
-    double endDistance = (endTime - intercept) / slope;
+    double endDist = startDist + timeDuration * averageSpeed;
 
-    if (slope > 1.0 || Double.isNaN(slope)) { // If the s/m is greater than 1 m/s (2.23694 mph) then assume the vehicle is stopped
+    if (averageSpeed < 1.0) { // If the m/s is smaller than 1 m/s (2.23694 mph) then assume the vehicle is stopped
       // Use small time increments instead of distance steps
       while (t < endTime) {
 
         t += 0.1;
         projection.add(new RoutePointStamped(d, 0, t)); // Assume that the vehicle is stationary
       }
     } else { // Vehicle is in motion so use distance steps
-      while (d < endDistance) {
+      while (d < endDist) {
 
         d += distanceStep;
-        projection.add(new RoutePointStamped(d, 0, d * slope + intercept));
+        projection.add(new RoutePointStamped(d, 0, d + (distanceStep / averageSpeed)));
```

Comment:
Shouldn't it be 
t + (distanceStep / averageSpeed)
and t needs to be updated as the loop runs

Also you could factor out distance / averageSpeed into a final variable called timeStep

Reply:
Ah. Good catch...